nixos/modules/services/backup/tarsnap.nix

```diff
@@ -242,9 +254,11 @@ in
 
     systemd.services."tarsnap@" = {
       description = "Tarsnap archive '%i'";
-      requires    = [ "network.target" ];
+      requires    = [ "network-online.target" ];
+      after       = [ "network-online.target" ];
 
-      path = [ pkgs.tarsnap pkgs.coreutils ];
+      path = [ pkgs.tarsnap pkgs.coreutils pkgs.iputils ];
+      preStart = "/run/current-system/sw/bin/sh -c 'while ! ping -q -c 1 betatest-server.tarsnap.com &> /dev/null; do true; done'";
```

Comment:
Yea, I wish the network-online.target stuff was enough but it doesn't seem to be and I'm not exactly sure why. The [docs](https://wiki.freedesktop.org/www/Software/systemd/NetworkTarget/) seem to suggest that network-online.target should simply work as is or you can use something like NetworkManager-wait-online.service/system-networkd-wait-online.service to force the **network.target** to wait until the network is online (like **network-online.target** is suppose to do) which is overkill/overreaching to enable. Now [systemd.networkd-wait-online.service had a change](https://github.com/systemd/systemd/commit/3005a221f1f2d7bc6e6f1a4df6b40e7d8bbd7e9e) about a year ago which removed the before requirement on network.target, so *that* would seem to suggest that enabling that service might actually make network-online.target work correctly now without requiring everything with dependencies on network.target to wait as well (didn't actually test this). The [NetworkManager service](http://cgit.freedesktop.org/NetworkManager/NetworkManager/tree/data/NetworkManager-wait-online.service.in#n6) still has a before requirement on network.target. We have [our own thing for NetworkManager](https://github.com/NixOS/nixpkgs/blob/6e6a96d42cf56cfcd042bbeab89e37f442f0cfcc/nixos/modules/services/networking/networkmanager.nix#L51) which triggers network-online.target when an interface comes up (which I don't know what exactly NetworkManager's definition of "up" is). Ultimately, if we have to enable a special service to get network-online.target to work for each possible network management system we support, some of them don't provide a built in facility for this waiting stuff (NetworkManager and systemd-networkd ship units to do this but seem to differ in their actual effect on the boot process as mentioned above) and they differ in their definition of what "online" means, then the persistent option is only going to work if users are using some limited set of network management systems and be broken (in the sense that tarsnap will bail out every time) otherwise. 

The PreStart hack ensures that the tarsnap only starts when we can actually reach out to the tarsnap server -- regardless of what is managing the network configuration -- since otherwise there's no point in performing a backup. Note, there is a timeout for starting a service (90s by default) so it is not going to sit there forever pinging away. The unit will just fail if we can't connect to the server after 90s, so if the system is extremely slow booting/waking up or the system is extremely slow getting a network connection, then the unit will fail even though it might have worked in a few minutes. We can of course tweak this with TimeoutStartSec in the service file and we could expose this to be configurable by the user if we think it needs to be. Definitely another downside to the PreStart thing, but it hasn't hit me in the year I've been doing it.

I'm happy to explore other options, I'm not very fond of the PreStart thing either, but it's the easiest way I know how to ensure that it works for nearly all cases. If we could get all our network stuff to agree on what network-online.target means, whether that is "I have an IP" or "I can connect to some arbitrary location" or something else, then we might have a clearer way forward, but I suspect they will probably not all mean "I can connect to the internet" since that is a fairly broad definition of "online". Thus, we are probably still going to need something which delays execution until we know that yes, the network is up/"online" *and* we've at least tried to connect to the outside world before failing.

And yes, persistent timers just track the last time the service was triggered (i.e., the whole service was asked to start, so regardless of it starts successfully or not) not the last time it completed successfully, which is fine for my uses.